### ISSUE TITLE
Bug: can not add excepted patter in no-restricted-imports rule

### ISSUE THREAD

**Issue by iserdmi** [eslint/eslint#16747]
### Environment

Node version: 18.12.1
npm version: 8.19.2 (but I use pnpm 7.21.0)
Local ESLint version: 8.31.0
Global ESLint version: —
Operating System: macOS Monterey 12.5

### What parser are you using?

Other

### What did you do?

My configuration:

```
rules:
  'no-restricted-imports':
    - error
    - patterns:
        - '@startup-fullstack/backend/*'
        - '!@startup-fullstack/backend/**/input'
```

```js
// Throw error (expected)
import { somethingPrivate } from '@startup-fullstack/backend/src/private'

// Throw error (not expected)
import { zCreateIdeaInput } from '@startup-fullstack/backend/src/trpc/createIdea/input'
import { zEditIdeaInput } from '@startup-fullstack/backend/src/trpc/editIdea/input'

```

### What did you expect to happen?

I expect, that it is possible to add excepted path pattern by adding "!..." in the pattern string

### What actually happened?

I am not able to import from the excepted path

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

Please, tell me, am I doing something wrong, or is it really a bug?

**Comment by mdjermanovic** [comment]
Hi @iserdmi, thanks for the issue!

The `patterns` option in this rule works per the [gitignore specification](https://git-scm.com/docs/gitignore). Patterns match both files and directories, and it is not possible to re-include a file if a parent directory of that file is excluded.

In your example, `'@startup-fullstack/backend/*` matches `@startup-fullstack/backend/src/` directory. Negated pattern `!@startup-fullstack/backend/**/input` does not match that directory, thus the directory stays restricted, and therefore all paths under the directory are considered restricted (even if they directly match the negated pattern), so this works as intended.

If you want to restrict all filepaths under `@startup-fullstack/backend/` except for those that end with `/input`, you can specify the list of patterns as follows: `['@startup-fullstack/backend/**', '!@startup-fullstack/backend/**/', '!@startup-fullstack/backend/**/input']`.

```
# restrict everything
@startup-fullstack/backend/**

# allow directories
!@startup-fullstack/backend/**/

# allow **/input
!@startup-fullstack/backend/**/input
```

[Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiaW1wb3J0IHsgc29tZXRoaW5nUHJpdmF0ZSB9IGZyb20gJ0BzdGFydHVwLWZ1bGxzdGFjay9iYWNrZW5kL3NyYy9wcml2YXRlJ1xuXG5pbXBvcnQgeyB6Q3JlYXRlSWRlYUlucHV0IH0gZnJvbSAnQHN0YXJ0dXAtZnVsbHN0YWNrL2JhY2tlbmQvc3JjL3RycGMvY3JlYXRlSWRlYS9pbnB1dCdcbmltcG9ydCB7IHpFZGl0SWRlYUlucHV0IH0gZnJvbSAnQHN0YXJ0dXAtZnVsbHN0YWNrL2JhY2tlbmQvc3JjL3RycGMvZWRpdElkZWEvaW5wdXQnIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby1yZXN0cmljdGVkLWltcG9ydHMiOlsiZXJyb3IiLHsicGF0dGVybnMiOlsiQHN0YXJ0dXAtZnVsbHN0YWNrL2JhY2tlbmQvKioiLCIhQHN0YXJ0dXAtZnVsbHN0YWNrL2JhY2tlbmQvKiovIiwiIUBzdGFydHVwLWZ1bGxzdGFjay9iYWNrZW5kLyoqL2lucHV0Il19XX0sImVudiI6eyJlczYiOnRydWV9fX0=)
